code/iaas/auth-logic/src/main/java/io/cattle/platform/iaas/api/auth/AbstractTokenUtil.java

```diff
@@ -396,4 +406,79 @@ public static boolean isUnrestrictedAccess(String accessMode){
         return UNRESTRICTED_ACCESSMODE.equalsIgnoreCase(accessMode);
     }
 
+    public Identity jsonToIdentity(Map<String, Object> jsonData) {
+        String externalId = ObjectUtils.toString(jsonData.get("externalId"));
+        String externalIdType = ObjectUtils.toString(jsonData.get("externalIdType"));
+        String name = ObjectUtils.toString(jsonData.get("name"));
+        String profilePicture = ObjectUtils.toString(jsonData.get("profilePicture"));
+        String profileUrl = ObjectUtils.toString(jsonData.get("profileUrl"));
+        String login = ObjectUtils.toString(jsonData.get("login"));
+        return new Identity(externalIdType, externalId, name, profileUrl, profilePicture, login);
+    }
+
+    public Token getUserIdentityFromJWT() {
+        Token token = new Token();
+        Map<String, Object> jsonData = getJsonData();
+        if (jsonData == null) {
+            return null;
+        }
+        AuthToken authToken= authTokenDao.getTokenByKey(getJWT());
+        String tokenId = (String)ApiContext.getContext().getIdFormatter().formatId(objectManager.getType(Account.class), authToken.getId());
+        token.setId(tokenId);
+        Object idObject = jsonData.get(USER_IDENTITY);
+        if (idObject != null) {
+            Map<String, Object> idMap = CollectionUtils.toMap(idObject);
+            Identity userIdentity = jsonToIdentity(idMap);
+            String userType = ObjectUtils.toString(jsonData.get(USER_TYPE), null);
+            token.setUserIdentity(userIdentity);
+            token.setUserType(userType);
+        }
+        return token;
+    }
+
+    public String getRedirectUrl() {
```

Comment:
getRedirectUrl() is calling the externalAuthService to get the redirect Url... this method is specific to external provider. Lets keep this with the ExternalProvider, since its does not apply to all providers.